### DISCUSSION TITLE
Consider tRPC for edge functions

### DISCUSSION THREAD

**Discussion by c00kiemon5ter** [supabase/supabase#7685] [Feature Requests]
Has anyone tried [tRPC](https://trpc.io/)? This seems to be a nice utility to use on edge functions along with [zod](https://zod.dev/).

**Comment by toddgeist** [comment]
zod and trpc are awesome.  I would love to see superbase typescript support adopt these two amazing libraries.

**Comment by Nikola-Milovic** [comment]
Any updates? Anyone managed to get it working?

**Comment by krisgrm** [comment]
Wondering the same

**Comment by agustif** [comment]
Maybe this docs from trpc help? https://trpc.io/docs/server/adapters/fetch

**Comment by krisgrm** [comment]
i was looking at this, so if i understand correctly I'd make 1 supabase edge function, wrap it with trpc router that would be it 

will report back

**Comment by iveshenry18** [comment]
Has anyone given this a try? It seems trivially doable, given tRPC has an [official Deno Deploy example](https://github.com/trpc/trpc/tree/main/examples/deno-deploy), but I'm curious if there are any Supabase-specific quirks that come up.

**Reply by iveshenry18** [reply]
(Updated 3/23/2024)

Was overly optimistic about "trivially," but I did manage to hack together an initial version of this. I had some challenges resolving the AppRouter type defined in the Deno function code in our app's react native frontend, so I ended up defining the tRPC router in its own npm library outside of both the client and server code, then importing that library into both the frontend code and the Deno function.

So the simplified directory structure is:
```
├── supabase
│   └── functions
|        └── function-1
├── function-libs
│   └── function-1
└── frontend
```
where `function-libs/function-1` is an npm package that exports both the appRouter and its type. Since Deno doesn't seem to support local npm packages, my current semi-hacky solution is to bundle this into the `dist` directory (`bun build ./index.ts --outfile=./dist/index.js --target=node`) and just import the bundled js into Deno. This setup doesn't get me type safety on the Deno side, but that should be doable by creating a `.d.ts` file with something like [bun-plugin-dts](https://github.com/wobsoriano/bun-plugin-dts).

Then in `supabase/functions/function-1/index.ts`:
```ts
import { fetchRequestHandler } from "@trpc/adapters/fetch";
import { serve } from "std/http/server.ts";
import type { Handler } from "std/http/server.ts";
import { appRouter } from "../../function-libs/function-1/dist/index.js";

const handler: Handler = (request) =>
  fetchRequestHandler({
    endpoint: "/trpc",
    req: request,
    router: appRouter,
    createContext: () => ({ 
      /* ...create context from env variables (API keys for other services, etc.) */ 
      const supabaseUrl = Deno.env.get("SUPABASE_URL");
      const supabaseAnonKey = Deno.env.get("SUPABASE_ANON_KEY");

      
      // To use the client's authorization, just forward the Authorization header from the request,
      // as described in https://supabase.com/docs/guides/functions/auth
      const authHeader = req.headers.get("Authorization");
      if (!authHeader) {
        throw new TRPCError({
          code: "BAD_REQUEST",
          message: "Authorization header missing",
        });
      }
      const supabaseClient = createClient<Database>(
        supabaseUrl,
        supabaseAnonKey,
        {
          global: { headers: { Authorization: authHeader } },
        }
      );
      
      // WARNING: this context object will not be type-safe and must be manually verified
      return { supabaseClient, /* other clients/keys */ }
    }),
  });

serve(handler);
```

I can then use middleware to create authenticated routes (`function-libs/function-1/index.ts`):

```ts
/* ... */
const authedProcedure = publicProcedure.use(async (opts) => {
  const currentUserRes = await opts.ctx.supabaseClient.auth.getUser();
  if (!currentUserRes.data.user) {
    throw new TRPCError({
      code: "UNAUTHORIZED",
      message: "User is not authenticated",
    });
  }
  return opts.next({
    ctx: {
      currentUser: currentUserRes.data.user,
    },
  });
})

/* ... */
```

In `frontend` I run `bun i ../function-libs/function-1` and import it as a normal npm package:
```ts
import { AppRouter } from "@my-scope/function-1";
import { createTRPCReact, httpBatchLink } from "@trpc/react-query";

const TRPC_URL = process.env.EXPO_PUBLIC_TRPC_URL;
if (TRPC_URL === undefined) {
  throw new Error("No tRPC URL defined");
}

export const trpc = createTRPCReact<AppRouter>();
export const createTrpcClient = (session: Session | null) => {
  return trpc.createClient({
    links: [
      httpBatchLink({
        url: TRPC_URL,
        headers: async () => ({
          "Content-Type": "application/json",
          Authorization: `Bearer ${
            (await supabase.auth.getSession()).data.session?.access_token ??
            supabaseAnonKey
          }`,
        }),
      }),
    ],
  });
};

```


Notable gotcha: I had to make sure I installed the _exact_ same trpc versions in all places, otherwise I was getting deep type errors. I may migrate to a monorepo to manage these dependencies.

**Reply by RyanClementsHax** [reply]
Thanks a ton for posting this. This was a real life saver.

**Reply by iveshenry18** [reply]
Glad it proved useful! I just updated it with more details and findings from deploying to production, so it should be a fairly complete solution now. Let me know if you have any questions or issues.

**Reply by RyanClementsHax** [reply]
Oh! Thanks for the update. I like how you grab the auth token for the trpc client. I'll steal that too 😄.

> // WARNING: this context object will not be type-safe and must be manually verified

You can actually make this type safe by doing something like this.

```ts
export const createContext = (opts: FetchCreateContextFnOptions) => {
    /* ...create context from env variables (API keys for other services, etc.) */
    const supabaseUrl = Deno.env.get("SUPABASE_URL");
    const supabaseAnonKey = Deno.env.get("SUPABASE_ANON_KEY");

    // To use the client's authorization, just forward the Authorization header from the request,
    // as described in https://supabase.com/docs/guides/functions/auth
    const authHeader = req.headers.get("Authorization");
    if (!authHeader) {
        throw new TRPCError({
            code: "BAD_REQUEST",
            message: "Authorization header missing",
        });
    }
    const supabaseClient = createClient<Database>(
        supabaseUrl,
        supabaseAnonKey,
        {
            global: { headers: { Authorization: authHeader } },
        }
    );

    return { supabaseClient /* other clients/keys */ };
};
export type Context = ReturnType<typeof createContext>;

export const t = initTRPC.context<Context>().create();
export const procedure = t.procedure.use(async ({ ctx }) => {
    // This is safe now
    ctx.supabaseClient
});
```

**Reply by DawidWraga** [reply]
⚠️Quote from supabase docs:

Never trust supabase.auth.getSession() inside server code such as middleware. It isn't guaranteed to revalidate the Auth token.

**Reply by iveshenry18** [reply]
for what it's worth, I ended up migrating to Hono, which provides similar type-safety but I found to be more feature-complete (supports multipart/form-data) and aligned with REST principles (normal paths rather than trpc's dot-style paths)

**Reply by RobSchilderr** [reply]
What about combining [tRPC with Hono](https://www.npmjs.com/package/@hono/trpc-server) on edge functions for better type safety and react query integration? seems interesting
